### PR TITLE
Refactor SQONBuilder to have type validation

### PR DESCRIPTION
**Pull request by joneubank**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,5 @@
 dist
 node_modules
 index.js
+
+.nyc_output/
```
Comment (justincorrigible) [inline .gitignore:6]: lol what?

File: src/index.ts [hunk 2]
```diff
@@ -1,3 +1,5 @@
-import SQON from './SQON';
+export * from './types/sqon';
+import { default as SQONBuilder } from './SQONBuilder';
+export { default as reduceSQON } from './utils/reduceSQON';
 
-export default SQON;
+export default SQONBuilder;
```
Comment (justincorrigible) [inline src/index.ts:5]: 👍

File: src/utils/cloneDeepPojo.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+/**
+ * Create a clone of the provided object, stripping out all functions.
+ * The resulting object should include a copy of the original object but containing only the objects, arrays,
+ * and primatives.
+ *
+ * This is used in the SQONBuilder to return a copy of the sqon data without passing along the builder functions.
+ * @param source
+ * @returns
+ */
+const cloneDeepPojo = <T extends object>(source: T): T => {
```
Comment (justincorrigible) [inline src/utils/cloneDeepPojo.ts]: POJSO ? lol

File: test/utils/filterDuplicates.spec.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import filterDuplicates from '../../src/utils/filterDuplicates';
+
+describe('filterDuplicates', () => {
+	it('no duplicates - not modified', () => {
+		const input = [1, 2, 3, 4];
+		const output = input.filter(filterDuplicates);
+		expect(output).deep.equal(input);
+	});
+	it('duplicates - removes duplicates', () => {
+		const input = [1, 1, 2, 3, 2, 3, 4, 4, 4, 4, 4];
```
Comment (justincorrigible) [inline test/utils/filterDuplicates.spec.ts:11]: 🤔 surely this works great, but the example looks a bit reductionist. no?

File: .gitignore (hunk 1 above)
Comment (joneubank) [reply .gitignore:6]: test coverage report

File: test/utils/filterDuplicates.spec.ts (hunk 4 above)
Comment (joneubank) [reply test/utils/filterDuplicates.spec.ts:11]: Open to other test case suggestions. The idea here is that this covers elements that are in mixed order and when there is more than 1 to remove.

File: src/utils/cloneDeepPojo.ts (hunk 3 above)
Comment (joneubank) [reply src/utils/cloneDeepPojo.ts]: yessir

**Pull request merged**